### PR TITLE
Enable VMR sync without passing GitHub/AzDO tokens

### PR DESCRIPTION
**Pull request by MilenaHristova**
https://github.com/dotnet/arcade/issues/10549

### CODE REVIEW

File: src/Microsoft.DotNet.Darc/src/Darc/Options/VirtualMonoRepo/VmrCommandLineOptions.cs [hunk 1]
```diff
@@ -24,17 +24,26 @@ internal abstract class VmrCommandLineOptions : CommandLineOptions
...
         var tmpPath = Path.GetFullPath(TmpPath ?? Path.GetTempPath());
-        var localDarcSettings = LocalSettings.LoadSettingsFile();
+        string gitHubToken = null;
+        string azureDevOpsToken = null;
+
+        try
+        {
+            var localDarcSettings = LocalSettings.LoadSettingsFile(this);
+            gitHubToken = GitHubPat ?? (localDarcSettings?.GitHubToken);
+            azureDevOpsToken = AzureDevOpsPat ?? (localDarcSettings?.AzureDevOpsToken);
```
Comment (premun) [inline src/Microsoft.DotNet.Darc/src/Darc/Options/VirtualMonoRepo/VmrCommandLineOptions.cs:34]:
```suggestion
            gitHubToken = GitHubPat ?? localDarcSettings?.GitHubToken;
            azureDevOpsToken = AzureDevOpsPat ?? localDarcSettings?.AzureDevOpsToken;
```

File: src/Microsoft.DotNet.Darc/src/Darc/Options/VirtualMonoRepo/VmrCommandLineOptions.cs [hunk 2]
```diff
@@ -24,17 +24,26 @@ internal abstract class VmrCommandLineOptions : CommandLineOptions
     public IServiceCollection RegisterServices()
     {
         var tmpPath = Path.GetFullPath(TmpPath ?? Path.GetTempPath());
-        var localDarcSettings = LocalSettings.LoadSettingsFile();
+        string gitHubToken = null;
+        string azureDevOpsToken = null;
+
+        try
+        {
+            var localDarcSettings = LocalSettings.LoadSettingsFile(this);
```
Comment (premun) [inline src/Microsoft.DotNet.Darc/src/Darc/Options/VirtualMonoRepo/VmrCommandLineOptions.cs:32]: You should try to put only the lines that you suspect are throwing in the `try {}`. So here it is just this line.

File: src/Microsoft.DotNet.Darc/src/Darc/Options/VirtualMonoRepo/VmrCommandLineOptions.cs [hunk 3]
```diff
@@ -24,17 +24,26 @@ internal abstract class VmrCommandLineOptions : CommandLineOptions
...
+            gitHubToken = GitHubPat ?? (localDarcSettings?.GitHubToken);
+            azureDevOpsToken = AzureDevOpsPat ?? (localDarcSettings?.AzureDevOpsToken);
+        }
+        catch (DarcException)
+        {
+
```
Comment (premun) [inline src/Microsoft.DotNet.Darc/src/Darc/Options/VirtualMonoRepo/VmrCommandLineOptions.cs:38]: Please put a comment inside of this catch block saying something like we want to allow null

File: src/Microsoft.DotNet.Darc/src/Darc/Options/VirtualMonoRepo/VmrCommandLineOptions.cs [hunk 4]
```diff
@@ -24,17 +24,26 @@ internal abstract class VmrCommandLineOptions : CommandLineOptions
...
+        }
 
         var services = new ServiceCollection();
 
         services.TryAddSingleton<IRemoteFactory>(_ => new RemoteFactory(this));
 
         services.AddVmrManagers(GitLocation, VmrPath, tmpPath, configure: sp =>
```
Comment (premun) [inline src/Microsoft.DotNet.Darc/src/Darc/Options/VirtualMonoRepo/VmrCommandLineOptions.cs:45]: Can you use the other overload that sends tokens directly?

File: src/Microsoft.DotNet.Darc/src/Darc/Options/VirtualMonoRepo/VmrCommandLineOptions.cs [hunk 5]
```diff
@@ -26,26 +26,25 @@ public IServiceCollection RegisterServices()
...
         string gitHubToken = null;
         string azureDevOpsToken = null;
+        LocalSettings localDarcSettings = null;
 
         try
         {
-            var localDarcSettings = LocalSettings.LoadSettingsFile(this);
-            gitHubToken = GitHubPat ?? localDarcSettings?.GitHubToken;
-            azureDevOpsToken = AzureDevOpsPat ?? localDarcSettings?.AzureDevOpsToken;
+            localDarcSettings = LocalSettings.LoadSettingsFile(this);
```
Comment (premun) [inline src/Microsoft.DotNet.Darc/src/Darc/Options/VirtualMonoRepo/VmrCommandLineOptions.cs:33]: I am thinking - we should only try to load the local settings if both tokens are null so we are not trying to read it always for no reason.

File: src/Microsoft.DotNet.Darc/src/Darc/Options/VirtualMonoRepo/VmrCommandLineOptions.cs [hunk 6]
```diff
@@ -26,26 +26,25 @@ public IServiceCollection RegisterServices()
...
         }
         catch (DarcException)
         {
-
+            // we want to allow null values for GitHubToken and AzureDevOpsToken
```
Comment (premun) [inline src/Microsoft.DotNet.Darc/src/Darc/Options/VirtualMonoRepo/VmrCommandLineOptions.cs:37]: @MilenaHristova my 2 cents - when adding comments, try to put in them what is not in the code around. The code quite clearly reads as "we don't need to have the tokens" so it's better to explain why (which can have some non-code related justification).

```suggestion
            // The VMR synchronization often works with public repositories where tokens are not required
```

File: src/Microsoft.DotNet.Darc/src/Darc/Options/VirtualMonoRepo/VmrCommandLineOptions.cs [hunk 7]
```diff
@@ -26,26 +26,25 @@ public IServiceCollection RegisterServices()
...
-            azureDevOpsToken = AzureDevOpsPat ?? localDarcSettings?.AzureDevOpsToken;
+            localDarcSettings = LocalSettings.LoadSettingsFile(this);
         }
         catch (DarcException)
         {
-
+            // we want to allow null values for GitHubToken and AzureDevOpsToken 
         }
 
+        gitHubToken = GitHubPat ?? localDarcSettings?.GitHubToken;
```
Comment (premun) [inline src/Microsoft.DotNet.Darc/src/Darc/Options/VirtualMonoRepo/VmrCommandLineOptions.cs:40]: Let's merge the initialization with the declaration

**Review by premun — CHANGES_REQUESTED** [review]
Please also remove the `darc authenticate` call from `test.yml`

File: src/Microsoft.DotNet.Darc/src/Darc/Options/VirtualMonoRepo/VmrCommandLineOptions.cs (hunk 5 above)
Comment (MilenaHristova) [reply src/Microsoft.DotNet.Darc/src/Darc/Options/VirtualMonoRepo/VmrCommandLineOptions.cs:33]: what if only one is null and we can read the other one from the settings?

Comment (premun) [reply src/Microsoft.DotNet.Darc/src/Darc/Options/VirtualMonoRepo/VmrCommandLineOptions.cs:33]: Then we should read it from the settings

**Review by premun — APPROVED** [review]

**Pull request merged**